tests/functional/r/return_in_finally.py

```diff
@@ -0,0 +1,41 @@
+# pylint: disable=missing-docstring,too-few-public-methods,useless-return,lost-exception,inconsistent-return-statements
+
+def second_favorite():
+    fruits = ["apple"]
+    try:
+        return fruits[1]
+    finally:
+        return fruits[0]  # [return-in-finally]
+
+
+def all_favorites():
+    fruits = ["orange", "kiwi", "pineapple"]
+
+    for fruit in fruits:
+        try:
+            return fruit
+        finally:
+            return  # [return-in-finally]
+
+
+def more_favorites():
+    fruits = ["orange", "kiwi", "pineapple"]
+
+    for fruit in fruits:
+        try:
+            return fruit
+        finally:
+            if len(fruit) > 7:
+                return "too many fruits"  # [return-in-finally]
+
+
+def even_more_favorites():
```

Comment:
What about an example with `return` in the finally block but lacks a `return` in the try block. (And would you consider that something that should still emit a message?)

Reply:
I'm imagining a legit use case like:

```python
def divide(numerator, denominator):
    try:
        result = numerator / denominator
    except ZeroDivisionError:
        try:
            send_error()
        finally:
            return
    return result
```

Reply:
A `return` in a `finally` will _always_ overwrite the return value. It doesn't matter if there's a `return` inside the `try` itself.

So even if your function catches an error and continues without raising any other exceptions, it will execute the `finally` and if there was a return statement, it will return this value and discard whatever was returned afterwards.

```py
def foo():
    try:
        raise Exception
    finally:
        return 1

    return 2

print(foo())
```

This prints `1`.

Reply:
Wait no. Scrap that (ignore my example).

Let me think about it for a bit.

Reply:
But in my example, only if the inner `try` executes, right?
```python
>>> divide(1, 4)
0.25
```

I think I'm asking what you think of checking whether or not the try/finally in question is guarded under `except`.

Reply:
Do you have any real world examples? If I see something like this in a MR/PR, I'd question if there's something wrong in general here. For example, the `send_error()` method should probably take care of any errors. So it would look like this:

```py
def divide(numerator, denominator):
    result = None
    try:
        result = numerator / denominator
    except ZeroDivisionError:
        send_error()  # send_error should handle its own exceptions
    return result
```

So I can't really think of a situation where this would make sense to have a `return` in a `finally`, let alone a `try` in an `except` clause.

Reply:
I think the most realistic case to consider would be when you're not in control of `send_error()`,  or if it's more idiomatic for send_error to raise documented, idiomatic exceptions like NotAuthorized or Http500Exception or whatnot. But I don't think we need to block a merge over this; a user in this scenario can just disable `return-in-finally` since they are doing something a little brittle.

Reply:
I know we're drifting off-topic now, but in such a case I would just have a module that takes care of external stuff I can't handle and call my "proxy"-methods instead.

I know it's not a thorough explanation of my thought process though and there still might be edge cases where one has to live with such a situation.